### PR TITLE
Improved soca ctests

### LINKED ISSUE #103: compare.sh tests failing on different machines
Tests relying on compare.sh are not bit reproducible and so are failing on some machines. Several things need to be looked at

1. I think there are still some bugs hidden in the code. E.g I should not be getting completely different answers in `soca_test_increment` depending on whether i run in Release or Debug
2. compiler flags for optimization and math model choice should be played with. Intel compilers are especially notorious for changing answers with some of the default optimization settings.
3. Add tolerances to the compare.sh script

### LINKED ISSUE #177: use the compare.py script for ctests
... I'm tired ctests failing on some machines due to slightly different answers

### PR DESCRIPTION
**Pull request by travissluka**
Two overall improvements to the SOCA ctests
1) everything that had been using `compare.sh` now uses `compare.py` which allows for tolerances.  There are now two tests for each test before, a run and a compare. The tolerances were made just large enough so that ctests pass on Discover with gcc and intel17. 
2) test dependencies are now specified which mean we can run tests in parallel using `ctest -j <n>` (On my 8 core desktop soca ctests only take 9 seconds now!)

closes #177
closes #103

Details:
* Discover passes the tests in both debug and release for gcc, but intel/debug crashes at `test_soca_enspert`. (deal with that another time)
* for any tests using the "run executable, compare log results" paradigm, for simplicity of the `CMakeLists.txt` file the `soca_add_test()` function should be used now instead of `ecbuild_add_test()`
* `soca_add_test()` assumes the test name suffix, input config file base name, and reference answers file base name, are all the same name, which should help slightly with test organization
* Tolerances for the compare test are specified in the `CMakeLists.txt` file, but can be overridden with the `-DSOCA_TESTS_FORC_DEFAULT_TOL=ON` CMake option. We'll probably want to use this to force tight tolerances on the Travis-CI builds so that more subtle changes to answers will be picked up on.
* `compare.py` is currently unable to handle the output from the `enshofx` test, so the compare portion of that test is disabled for now. To be fixed in a later PR to oops.
* some of the tolerances needed seem a little large (e.g. 2e-2 needed for `3dvar_godas`). Hopefully in the future we'll discover compiler flags or bugs in the code that will let us tighten these up a bit.

### CODE REVIEW

**Comment by travissluka** [comment]
@guillaumevernieres @ShastriPaturi If you can confirm that all the soca ctests pass on Theia

**Comment by ShastriPaturi** [comment]
The following tests FAILED:
	  3 - test_soca_forecast_identity_compare (Failed)
	  5 - test_soca_forecast_identity_cice_compare (Failed)
	  7 - test_soca_forecast_mom6_compare (Failed)
	 10 - test_soca_enspert_compare (Failed)
	 29 - test_soca_dirac_socahyb_cov_compare (Failed)
	 31 - test_soca_hofx3d_compare (Failed)
	 33 - test_soca_hofx_compare (Failed)
	 36 - test_soca_3dvar_soca_compare (Failed)
	 38 - test_soca_3dvar_godas_compare (Failed)
	 40 - test_soca_3dvarfgat_compare (Failed)
	 42 - test_soca_3dhyb_compare (Failed)
	 44 - test_soca_3dhybfgat_compare (Failed)
	 46 - test_soca_checkpointmodel_compare (Failed)

If these tests are supposed to fail, then I am ok.

**Comment by travissluka** [comment]
@ShastriPaturi no, all the tests should be passing now

**Comment by travissluka** [comment]
@ShastriPaturi this relies on a change just made to oops/develop a couple of hours ago. make sure you're using develop and not release/soca-stable-nightly

**Comment by ShastriPaturi** [comment]
> @ShastriPaturi this relies on a change just made to oops/develop a couple of hours ago. make sure you're using develop and not release/soca-stable-nightly

Thanks for the clarification.
Let me switch

**Comment by guillaumevernieres** [comment]
> @guillaumevernieres @ShastriPaturi If you can confirm that all the soca ctests pass on Theia

:+1:  @travissluka All good on Theia and singularity

File: test/CMakeLists.txt [hunk 1]
```diff
@@ -31,32 +30,31 @@ list( APPEND soca_test_input
   testinput/parameters_bump_cov_nicas.yml
   testinput/parameters_bump_loc.yml
   testinput/state_test.yml
-  testinput/static_SocaError_init.yml
-  testinput/static_SocaError_test.yml
+  testinput/static_error_init.yml
+  testinput/static_error_test.yml
```
Comment (guillaumevernieres) [inline test/CMakeLists.txt]: Why the name change? SocaError is the Covariance being initialized and tested. I'm OK with the lower case, but I'd rather keep socaerror. We might have to initialize/test something else in the future.

**Comment by ShastriPaturi** [comment]
> @guillaumevernieres @ShastriPaturi If you can confirm that all the soca ctests pass on Theia

@travissluka everything good on theia

**Comment by markjolah** [comment]
All tests are now passing for me and run fine in parallel (gnu-9.1 / openmpi 4.0.1).  Previous to this PR I was getting multiple failures due to tolerance issues.  Thanks @travissluka, this will make it easier for me to check other JEDI changes don't break SOCA.

**Comment by guillaumevernieres** [comment]
@travissluka , very nice! Just have 1 comment.

> All tests are now passing for me and run fine in parallel (gnu-9.1 / openmpi 4.0.1). Previous to this PR I was getting multiple failures due to tolerance issues. Thanks @travissluka, this will make it easier for me to check other JEDI changes don't break SOCA.

:+1:

**Review by guillaumevernieres — APPROVED** [review]

**Review by aerorahul — APPROVED** [review]
Code changes are good.
All tests (except `test_soca_3dvar_soca_compare` and `test_soca_3dhyb_compare`) pass on my Mac.
Tolerances are a tad bit small.
Thanks @travissluka!

**Review by ShastriPaturi — APPROVED** [review]

**Comment by travissluka** [comment]
@aerorahul feel free to increase the tolerances on those tests before we merge. How much bigger does it need to be?

**Comment by aerorahul** [comment]
@travissluka here are the tolerances for which the tests pass:

For 37: test_soca_3dvar_soca_compare, passing tolerance is 3.0e-4. It is currently set to 3.0e-5
Also, can you rename that test to `test_soca_3dvar_compare` and the associated test to `test_soca_3dvar`. I know soca has its own BE, but this is known; isn't it!

For 42: test_soca_3dhyb_compare, passing tolerance is 5.0e-4. It is currently set to 4.0e-4

**Comment by travissluka** [comment]
@aerorahul since @guillaumevernieres already merged I'll let you go ahead adjust the tolerances in a separate PR

**Pull request merged**